src/lightning/data/cache/dataloader.py

```diff
@@ -0,0 +1,196 @@
+        return torch.equal(d1, d2)
+    return d1 == d2
+
+
+class CacheDataset(Dataset):
+    def __init__(
+        self, dataset: Dataset, cache_dir: Optional[str], chunk_size: Optional[int], compression: Optional[str]
+    ):
+        self._datataset = dataset
+        if cache_dir is None:
+            cache_dir = os.path.join(os.getcwd(), "cache_dir", datetime.now().strftime("%m-%d-%Y-%H-%M"))
+            os.makedirs(cache_dir, exist_ok=True)
+            chunk_size = 2 << 26
+        self.cache = Cache(cache_dir, chunk_size=chunk_size, compression=compression)
+        self.is_deterministic = False
+
+    def __len__(self) -> int:
+        return len(self.cache) if self.cache.filled else len(self._datataset)
+
+    def __getitem__(self, index):
+        data_1 = self.cache[index] if self.cache.filled else self._datataset[index]
+        if not self.cache.filled:
+            if not self.is_deterministic:
+                data2 = self._datataset[index]
+                if not _equal_items(data_1, data2):
+                    raise ValueError(
+                        f"Your dataset items aren't deterministic. Found {data_1} and {data2} for index {index}."
+                        " HINT: Use the `lightning.data.cache.Cache` directly within your dataset."
+                    )
+                self.is_deterministic = True
+            self.cache[index] = data_1
+        return data_1
+
+
+class CacheCollateFn:
+    def __init__(self):
+        self.collate_fn = default_collate
+
+    def __call__(self, items):
+        if all(item is None for item in items):
+            return None
+        return self.collate_fn(items)
+
+
+class _SingleProcessDataLoaderIterPatch(_SingleProcessDataLoaderIter):
+    """This is overriden to inform the cache is done chunking."""
+
+    def _next_data(self):
+        try:
+            return super()._next_data()
+        except StopIteration:
+            for v in self._dataset_fetcher.dataset.__dict__.values():
+                if isinstance(v, Cache):
+                    v.done()
+            raise StopIteration()
+
+
+class _MultiProcessingDataLoaderIterPatch(_MultiProcessingDataLoaderIter):
+    def __init__(self, loader):
+        # Patch PyTorch worker loop
+        from torch.utils.data._utils import worker
+
+        from lightning.data.cache.worker import _worker_loop
+
+        worker._worker_loop = _worker_loop
+        super().__init__(loader)
+
+
+class LightningDataLoader(DataLoader):
+    def __init__(
+        self,
+        dataset,
+        *args,
+        sampler=None,
+        batch_sampler=None,
+        num_workers=0,
+        shuffle: bool = False,
+        generator=None,
+        batch_size=1,
+        drop_last=False,
+        cache_dir: Optional[str] = None,
+        chunk_size: Optional[int] = 2 << 26,
+        compression: Optional[str] = None,
+        **kwargs,
+    ):
+        if sampler:
+            raise Exception(
```

Comment:
```suggestion
            raise ValueError(
```

never raise a bare exception ;)